include/RAJA/policy/cuda/MemUtils_CUDA.hpp

```diff
@@ -299,184 +299,207 @@ cudaDeviceProp& device_prop()
 
 struct CudaFixedMaxBlocksData
 {
-  int multiProcessorCount;
-  int maxThreadsPerMultiProcessor;
+  int device_sm_per_device;
+  int device_max_threads_per_sm;
 };
 
 RAJA_INLINE
-size_t cuda_max_blocks(size_t block_size)
+CudaFixedMaxBlocksData cuda_max_blocks()
 {
-  static CudaFixedMaxBlocksData data = []() {
-    cudaDeviceProp& prop = cuda::device_prop();
-    return CudaFixedMaxBlocksData{prop.multiProcessorCount,
-                                  prop.maxThreadsPerMultiProcessor};
-  }();
+  static thread_local CudaFixedMaxBlocksData data {
+      cuda::device_prop().multiProcessorCount,
+      cuda::device_prop().maxThreadsPerMultiProcessor };
 
-  size_t max_blocks = data.multiProcessorCount *
-                  (data.maxThreadsPerMultiProcessor / block_size);
-
-  return max_blocks;
+  return data;
 }
 
 struct CudaOccMaxBlocksThreadsData
 {
-  size_t prev_shmem_size;
-  int max_blocks;
-  int max_threads;
+  size_t func_dynamic_shmem_per_block;
+  int func_max_blocks_per_device;
+  int func_max_threads_per_block;
 };
 
-template < typename RAJA_UNUSED_ARG(UniqueMarker), typename Func >
+template < typename RAJA_UNUSED_ARG(UniqueMarker) >
 RAJA_INLINE
-void cuda_occupancy_max_blocks_threads(Func&& func, size_t shmem_size,
-                                       int &max_blocks, int &max_threads)
+CudaOccMaxBlocksThreadsData cuda_occupancy_max_blocks_threads(const void* func,
+    size_t func_dynamic_shmem_per_block)
 {
-  static constexpr int uninitialized = -1;
-  static constexpr size_t uninitialized_size_t = std::numeric_limits<size_t>::max();
-  static thread_local CudaOccMaxBlocksThreadsData data = {
-      uninitialized_size_t, uninitialized, uninitialized};
+  static thread_local CudaOccMaxBlocksThreadsData data {
```

Comment:
Does '-1' indicate an unspecified value here and in several places below?  If so, would be more clear to use a macro or constexpr variable named `unspecified` (for example) rather than a magic number?

Reply:
It used to be named, I'll add that back in.